tasking/pulp/tasking/base.py

```diff
@@ -0,0 +1,444 @@
+        :type args:     list
+
+        :param kwargs:  Original keyword arguments for the executed task.
+        :type kwargs:   dict
+
+        :param einfo:   celery's ExceptionInfo instance, containing serialized traceback.
+        :type einfo:    ???
+        """
+        if isinstance(exc, PulpCodedException):
+            _logger.error(_('Task failed : [%(task_id)s] : %(msg)s') %
+                         {'task_id': task_id, 'msg': str(exc)})
+            _logger.error(traceback.format_exc())
+        else:
+            _logger.error(_('Task failed : [%s]') % task_id)
+            # celery will log the traceback
+
+        if not self.request.called_directly:
+            task_status = TaskStatus.objects.get(pk=task_id)
+            task_status.state = TaskStatus.FAILED
+            task_status.finished_at = timezone.now()
+            task_status.result = {'traceback': einfo.traceback}
+            if not isinstance(exc, PulpException):
+                exc = PulpException(str(exc))
+                task_status.result = exc.to_dict()
+
+            task_status.save()
+
+    def _get_parent_arg(self):
+        """Return a dictionary with the parent set if running inside of a Task"""
+        parent_arg = {}
+        current_task_id = get_current_task_id()
+        if current_task_id is not None:
+            current_task_obj = TaskStatus.objects.get(current_task_id)
+            parent_arg['parent'] = current_task_obj
+        return parent_arg
+
+
+def cancel(task_id):
+    """
+    Cancel the task that is represented by the given task_id. This method cancels only the task
+    with given task_id, not the spawned tasks. This also updates task's state to 'canceled'.
+
+    :param task_id: The ID of the task you wish to cancel
+    :type  task_id: basestring
+
+    :raises MissingResource: if a task with given task_id does not exist
+    :raises PulpCodedException: if given task is already in a complete state
+    """
+    try:
+        task_status = TaskStatus.objects.get(pk=task_id)
+    except TaskStatus.DoesNotExist:
+        raise MissingResource(task_id)
+
+    if task_status.state in TASK_FINAL_STATES:
+        # If the task is already done, just stop
+        msg = _('Task [%(task_id)s] already in a completed state: %(state)s')
+        _logger.info(msg % {'task_id': task_id, 'state': task_status.state})
+        return
+
+    celery_controller.revoke(task_id, terminate=True)
+    task_status.state = TaskStatus.CANCELED
+    task_status.save()
+
+    msg = _('Task canceled: %(task_id)s.')
+    msg = msg % {'task_id': task_id}
+    _logger.info(msg)
+
+
+def get_current_task_id():
+    """"
+    Get the current task id from celery. If this is called outside of a running
+    celery task it will return None
+
+    :return: The ID of the currently running celery task or None if not in a task
+    :rtype: str
+    """
+    if current_task and current_task.request and current_task.request.id:
```

Comment:
A nice opportunity for my favorite contextmanager! https://docs.python.org/3/library/contextlib.html#contextlib.suppress

```
with supress(AttributeError):
    return current_task.request.id
```